ext/ClimaCommsCUDAExt.jl

```diff
@@ -50,4 +50,63 @@ ClimaComms.elapsed(f::F, ::CUDADevice, args...; kwargs...) where {F} =
 ClimaComms.assert(::CUDADevice, cond::C, text::T) where {C, T} =
     isnothing(text) ? (CUDA.@cuassert cond()) : (CUDA.@cuassert cond() text())
 
+cuda_thread_index_1D() =
+    (CUDA.blockIdx().x - 1) * CUDA.blockDim().x + CUDA.threadIdx().x
+
+function cuda_device_limits_1D()
+    cuda_device = CUDA.device()
+    max_blocks =
+        CUDA.attribute(cuda_device, CUDA.DEVICE_ATTRIBUTE_MAX_GRID_DIM_X)
+    max_threads_in_block =
+        CUDA.attribute(cuda_device, CUDA.DEVICE_ATTRIBUTE_MAX_THREADS_PER_BLOCK)
+    return (; max_blocks, max_threads_in_block)
+end
+
+function threaded(f::F, ::CUDADevice, ::Val, itr) where {F}
+    length(itr) > 0 || return nothing
+
+    Base.require_one_based_indexing(itr)
+    function call_f_from_thread()
+        item_index = cuda_thread_index_1D()
+        item_index <= length(itr) && @inbounds f(itr[item_index])
+        return nothing
+    end
+    kernel = CUDA.@cuda launch=false call_f_from_thread()
+    optimal_threads_in_block = CUDA.launch_configuration(kernel.fun).threads
+
+    (; max_blocks, max_threads_in_block) = cuda_device_limits_1D()
+    blocks = min(cld(length(itr), optimal_threads_in_block), max_blocks)
+    threads_in_block = min(cld(length(itr), blocks), max_threads_in_block)
+    length(itr) <= threads_in_block * blocks ||
+        return threaded(f, CUDADevice(), 2, itr) # Coarsen threads if necessary.
+
+    CUDA.@sync kernel(; threads = threads_in_block, blocks)
+end
+
+function threaded(f::F, ::CUDADevice, min_items_in_thread::Int, itr) where {F}
+    length(itr) > 0 || return nothing
+    min_items_in_thread > 0 ||
+        throw(ArgumentError("`coarsen` value must positive"))
+
+    Base.require_one_based_indexing(itr)
+    function call_f_from_thread(items_in_thread)
+        thread_index = cuda_thread_index_1D()
+        first_item_index = items_in_thread * (thread_index - 1) + 1
+        last_item_index = items_in_thread * thread_index
```

Comment:
This may not always be the best way to iterate with thread coarsening. For example, one option is to use the following approach:
(pseudocode)
```
for i in thread_st:offset:thread_en
    process...f[i]
end
```
where `offset` is determined by the number of threads-per-block.